sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexComplexityEstimator.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.spark.rapids
+
+import org.apache.spark.sql.types.DataTypes
+
+object RegexComplexityEstimator {
+  private def countStates(regex: RegexAST): Int = {
+    regex match {
+      case RegexSequence(parts) =>
+        parts.map(countStates).sum
+      case RegexGroup(true, term) =>
+        1 + countStates(term)
+      case RegexGroup(false, term) =>
+        countStates(term)
+      case RegexCharacterClass(_, _) =>
+        1
+      case RegexChoice(left, right) =>
+        countStates(left) + countStates(right)
+      case RegexRepetition(term, QuantifierFixedLength(length)) =>
+        length * countStates(term)
+      case RegexRepetition(term, SimpleQuantifier(ch)) =>
+        ch match {
+          case '*' =>
+            countStates(term)
```

Comment:
I think for `*`, you still need at least 1 state to handle the case where it can be consuming input. You actually have the 1 state for `?`, which is the same thing

Reply:
For something like `x*` where `x` is any pattern, we only need the states for `x`. If we consume `x` once or more, we just transition the state back to itself. If we consume `x` 0 times, we  would have a transition to skip the state entirely. Either way, I think the number of states for `x*` should be equal to the number of states for `x`, which is `countStates(x)`.

For `?`, cuDF treats `x?` as `x|NOP` so the extra +1 is for the NOP instruction. 

I explain this in a bit more detail in the design doc.